Improve error handling and observability

Veetaha opened the pull request:
> There may be typos in the example screenshots. Unfortunately, I made them before I used `typos` to fix them. So if you see any typos in the examples here they may already be fixed.

Closes https://github.com/rust-marker/marker/issues/169

Refactored error handling and observability

Added a new crate called `marker_error` that is reused between `cargo-marker`, `marker_adapter`, and `marker_rustc_driver` to handle errors. The documentation in the crate and the crate's README.md should tell for itself.

# Error handling

In short, there is a generic error type `Error<Kind = Never>` with an optional `Kind` type parameter. It allows for both dynamic string-based errors and structural error handling by matching on a `Kind` enum. It is backed by [`thiserror`](https://docs.rs/thiserror) for implementing the `Error` trait and [`miette`](https://docs.rs/miette) for implementing the error reporting providing additional features like help and (not used yet) code snippets and span highlighting.

From code organization POV, there should generally be just one `ErrorKind` type per crate. And that type should live in `{crate}/src/error.rs` file. There is not much benefit from defining many error enums in different sets of operations. Of course, the latter approach is more convenient for structural error handling when you need to exhaustively match on all potential error kinds, but usually you don't need to do that and a single `ErrorKind` per crate is sufficient.

## Example errors

### Missing lints

<details>
<summary>Details</summary>


<img width="664" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/82cb95b6-febe-4106-b621-bcce444d6e45">

</details>

### Missing driver

<details>
<summary>Details</summary>


<img width="1044" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/ebec7aba-887a-4b2e-abd9-659c68ba966f">

</details>

### Caught a deny lint

<details>
<summary>Details</summary>


<img width="693" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/e2defd60-8f62-4460-bffb-056512872b26">

</details>

### Invalid path to a lint crate

<details>
<summary>Details</summary>


<img width="722" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/782765c9-0705-484e-a545-a84217bb677b">

</details>

### Invalid lint crate from the registry

<details>
<summary>Details</summary>


<img width="860" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/5b9b3c84-ac11-490d-a226-f8b83a2058df">

</details>

### Invalid config schema in TOML

<details>
<summary>Details</summary>


<img width="573" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/556bd506-f0ad-4069-a0bd-f8e33324748c">

</details>

We now also validate for unknown fields in the config

<details>
<summary>Details</summary>


<img width="552" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/bfde2127-ba5f-4839-8eca-722deeb5bf2d">

</details>


### Couldn't find the Cargo.toml

<details>
<summary>Details</summary>


<img width="753" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/0da0f1db-878b-4d29-b83b-a26fbcaef9f3">

</details>

### Failure to find the driver

<details>
<summary>Details</summary>


<img width="769" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/032c201c-2209-452a-b490-c3793610430d">

</details>

### Toolchain is not installed

<details>
<summary>Details</summary>


<img width="1026" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/23262376-a884-48d9-ab70-d4fa8c06d9c2">

</details>

# Panic handling

Tested it on this code.
```rust
let _span = info_span!("span_name", key = "value", "message").entered();

panic!("Intentional panic");
```

There is a custom panic hook in `cargo-marker` that renders the panic message and provides users with the link to report the panic. One of the links is at the top of the output. If the user's terminal supports it then it will be displayed as a clickable link (internal_marker_panic), and there is a textual link at the bottom. The backtrace is automatically captured and displayed on panic regardless of `RUST_BACKTRACE` env var. The span trace is not captured by default.

<details>
<summary>Details</summary>


<img width="857" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/10fc512a-3cc2-4868-b4b6-931ad36dd705">

</details>

To capture the spantrace the user needs to enable the logging with `MARKER_LOG` env variable.

<details>
<summary>Details</summary>


<img width="936" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/b0303e56-26cd-414e-982a-eff6bdec79da">

</details>

# Debugging error handling

Unlike with panic regular errors display neither the backtrace nor the spantrace. It's possible to force capturing both of them with `MARKER_ERROR_TRACE=1` env variable.

<details>
<summary>Details</summary>


<img width="965" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/7fae2326-84ae-49dd-92b5-2f8de450e0df">

</details>

To be able to capture the spantrace the `MARKER_LOG` env var needs to also be set just like with panics.

# Error handling in `marker_rustc_driver`

Error handling in `marker_rustc_driver` is hard. I tried to improve it where possible but its mostly limited by the rustc APIs that are used there. For example, the version mismatch between the marker API used in the lint crate and in the driver is still reported as a panic, but for example, a mistake in `MARKER_LINT_CRATES` is reported more nicely.

<details>
<summary>Details</summary>


<img width="1230" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/dbddaf1d-54ac-4641-b4f3-1922fac7cbf2">

</details>

I think the status of error handling in `marker_rustc_driver` can be improved, but not in this PR. This PR only added a framework for that and made `cargo-marker` as its first and main user.

# Observability

Instrumented the code with [`tracing`](https://docs.rs/tracing). Removed the `verbose` boolean threaded through functions here and there. Added logging of executed commands. It's possible to enable verbose logging using the [`EnvFilter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html) syntax with the `MARKER_LOG` env variable.

The `MARKER_LOG` env variable is responsible for configuring the logging only in `cargo-marker`. The `marker_rustc_driver` has its own `tracing` logging setup which we can't easily change. Rustc uses the `rustc_log` crate to set its own implementation of the `tracing` subscriber. Unfortunately, this means we can't add the [`ErrorLayer`](https://docs.rs/tracing-error/latest/tracing_error/struct.ErrorLayer.html) there, because we can't modify that code, so there will be no spantrace capturing in `marker_rustc_driver`.

Also used the `yansi` crate for colored output. This crate is preparing for `1.0` release and doesn't have an ultimate `1.0` version, but it has `1.0.0-rc.1`, so I used that. I selected this crate because it has a convenient syntax for adding ANSI style to messages and it seems to be one of the maintained coloring crates and it supports style wrapping.

All command invocations are captured in the logs with some simple home-made syntax highlighting applied to them. To log a command invocation you need to manually call the `.log()` method from the extension trait for `std::command::Command`. It isn't pretty and I'm thinking of making a wrapper type for `std::command::Command` to make logging fully automatic and also remove some repetitive code.

There is also a home-made syntax highlighting for TOML snippets. It doesn't do any robust TOML parsing but highlights just some simple patterns and punctuation. You may have noticed that in the error messages higher already.


# Example logs

<details>
<summary>Details</summary>


<img width="1234" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/4e69e84f-92f8-46ba-992f-2a35f8fe9030">

</details>


# User output

There is a distinction between logs and user output. Logs are always hidden by default (unless they are at the `warn!()` or `error!()` level). The user output is now more in line with the `cargo` style of output. We print the stages of linting with `Marker {stage description}` along with the other output from cargo.

<details>
<summary>Details</summary>


<img width="652" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/fe9b2c3b-a934-4e25-b2ce-acceaa5ab7d7">

</details>

xFrednet commented:
Thank you for the gigantic update. Due to its size, it'll take some time to do a full review. The example output already looks really cool IMO. You should get a full review withing a week :)

xFrednet commented on cargo-marker/src/observability/display.rs:
```diff
@@ -0,0 +1,47 @@
+/// Displays a stage of work that `cargo-marker` performs.
+/// This is specifically formatted to be aligned with the other cargo
+/// output like `Compiling` or `Checking`.
+pub(crate) fn print_stage(name: &str) {
+    println!("      {} {}", "Marker".bold().green(), stage(name));
+}
```
Optional suggestion, an empty line, before the stage, would make the output more readable for me:

```
      Marker compiling rustc driver
   Compiling rustc_tools_util v0.3.0
   Compiling libloading v0.8.0
   Compiling bumpalo v3.13.0
   Compiling marker_api v0.3.0-dev ([...]/marker/marker_api)
   Compiling marker_rustc_driver v0.3.0-dev ([...]/marker/marker_rustc_driver)
   Compiling marker_utils v0.3.0-dev ([...]/marker/marker_utils)
   Compiling marker_adapter v0.3.0-dev ([...]/marker/marker_adapter)
    Finished dev [unoptimized + debuginfo] target(s) in 7.72s

      Marker compiling lints
    Updating crates.io index
   Compiling proc-macro2 v1.0.66
   Compiling unicode-ident v1.0.11
   Compiling syn v1.0.109
   Compiling quote v1.0.33
   Compiling visibility v0.0.1
   Compiling marker_api v0.3.0-dev ([...]/marker/marker_api)
   Compiling marker_lints v0.3.0-dev ([...]/marker/marker_lints)
    Finished release [optimized] target(s) in 4.33s

      Marker linting
    Checking marker_api v0.3.0-dev ([...]/marker/marker_api)
   Compiling marker_rustc_driver v0.3.0-dev ([...]/marker/marker_rustc_driver)
    Checking marker_error v0.3.0-dev ([...]/marker/marker_error)
    Checking marker_uitest v0.3.0-dev ([...]/marker/marker_uitest)
```

```suggestion
pub(crate) fn print_stage(name: &str) {
    println!("\n      {} {}", "Marker".bold().green(), stage(name));
}
```

Veetaha commented on marker_error/README.md:
```diff
@@ -0,0 +1,140 @@
+
+Use `Result` type for reporting such errors. The crate that is responsible for rendering this type of errors was selected to be [`miette`]. This crate renders the chain of errors that happened during the execution of `marker` code in a readable colorful way. It also allows including the snippets of text that the error refers to; for example, if the config is invalid, we may show the snippet of th ...
+
+[`miette`] builds on top of [`thiserror`] and extends the `std::error::Error` with additional functionality like the one mentioned above, error codes, help messages, multiple error handling, etc. This crate propagates using both [`miette`] and `thiserror` in combination to achieve the goal of user-friendly error reporting.
```
Yeah, I had some difficulties when including this doc as a crate-level doc. I think `rustdoc` was complaining that I wasn't using an automatic link. I mistakenly put the link declaration in the wrong place at that time and thus was getting that error, so I removed it. Now I figured this out, and will fix, thanks

Veetaha commented on marker_error/src/trace.rs:
```diff
@@ -0,0 +1,68 @@
+                write!(f, "{}\n{spantrace}", "Spantrace:".red().bold())?;
+            },
+            tracing_error::SpanTraceStatus::EMPTY => {
+                write!(
+                    f,
+                    "{}. It may be expected if the error happened outside of \
+                    any spans, or those spans were not enabled. You may try \
+                    capturing it by enabling {} logging or increasing its level.",
+                    "⚠️  Spantrace wasn't captured".yellow(),
+                    "tracing".bold()
+                )?;
```
There is a dilemma here. This crate is used in several different binaries. Every such binary has its own env variable used for enabling the tracing. For example, `cargo-marker` uses `MARKER_LOG`, but `marker_rustc_driver` uses `RUSTC_LOG`, because it uses `rustc_log` crate to initialize `tracing`.
So if I were to hardcode `MARKER_LOG` in the error message here, then it would be wrong in the context of `RUSTC_LOG`. I considered using a global variable to deliver the info about the tracing env var used to this crate, but meh. 
The other compromise option is to write "Enable spantrace with `MARKER_LOG` for `cargo_marker` or `RUSTC_LOG` for `marker_rustc_driver`".

But... There is another caveat. Spantraces will never be captured for `marker_rustc_driver`. I haven't figured out how to do that yet. The problem is that `marker_rustc_driver` uses `rustc_log` crate that uses some custom internal `tracing` subscribers which I wouldn't like to override, but I also can't extend the `tracing` setup because `rustc_log` uses `tracing::subscriber::set_global_default()` [here](https://github.com/rust-lang/rust/blob/f6faef447536204549eb47f12884d3e229fe87d3/compiler/rustc_log/src/lib.rs#L95-L99). This isn't composable. once the tracing subscriber is it's possible to get it back, but it's not possible to extend it.

So even if I would say "Use `RUSTC_LOG` to enable spantrace capturing", then this wouldn't be true, because there isn't an `ErrorLayer` in the tracing subscriber for `marker_rustc_driver` and it won't work. 

It may be possible to modify `rustc_log` crate to expose the API that enables adding an `ErrorLayer` to the subscriber, but I didn't do it in this scope. I don't think it is critical, and may be done any time layer.

For now I can add "Use `MARKER_LOG` for cargo-marker; spantrace capturing in `marker_rustc_driver` is not supported yet"

Veetaha commented on marker_error/src/lib.rs:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    /// Make a root error that doesn't have a source error
+    pub fn root(message: impl Into<String>) -> Self {
```
Yes, there is a huge difference. `impl Into<String>`. The difference is that `into` consumes the value, but `to_string()` takes it by referece
```rust
fn into(self) -> String;
// vs
fn to_string(&self) -> String;
```
It means `impl Into<String>` doesn't do an allocation if you pass it an owned `String`, because such conversion would be noop. But if you pass `String` to `ToString`, it will always make a clone

Veetaha commented on cargo-marker/src/backend/toolchain.rs:
```diff
@@ -53,108 +55,114 @@ impl Toolchain {
-    let string_path = String::from_utf8(output.stdout).map_err(|err| ExitStatus::fatal(err, "incorrect bytes"))?;
-    let path = PathBuf::from_str(string_path.trim())
-        .map_err(|err| ExitStatus::fatal(err, format!("failed to parse path for `{tool}`")))?;
+    let string_path = String::from_utf8(output.stdout).context(|| "incorrect bytes")?;
+    let path = PathBuf::from_str(string_path.trim()).context(|| format!("failed to parse path for `{tool}`"))?;
```
Ah, in this case it's possible to use `PathBuf::from` instead of `from_str`, because the operation of converting from string to `PathBuf` is infallible.

Anyway, you are right, fixed the capitalization

Veetaha commented on marker_rustc_driver/src/main.rs:
```diff
@@ -228,97 +229,117 @@ fn main() {
-            println!("marker-api: {}", marker_api::MARKER_API_VERSION);
+            // This is a bit of a hack, but this way we can emit our own errors
+            // without having to change the rustc driver.
+            #[allow(deprecated)]
```
I didn't know about this, thanks. Even though I think it's better to use less unstable things, I see you have a different opinion at least in the context of `marker_rustc_driver`, I'll apply the suggestion

Veetaha replied:
Makes sense, that's cool 👍

Veetaha commented:
I addressed the review comments.

>  Rustc has the convention to always use a lower case start, but it felt like you preferred an upper case start. I'm fine with either option, I think consistency is the important part :)

Yeah, I'm used to write capitalized error messages. As for me, it's weird if the error message doesn't begin with a capital letter. Bceause sometimes I want to write several sentences or even whole paragraphs of text. It seems trage that such a paragraph would begin with a lower case letter.

I know rustc uses lower case convention for error messages and it separates sentences with `;`  instead. That seems unordinary for me. I don't have a strong opinion on this, and we may change this convention at any point, add a marker lint for that as well.
